### PR TITLE
Add Tizen support

### PR DESCRIPTION
**Pull request by JoonghyunCho**
This PR enables Tizen support to `Xamarin.Forms.PancakeView`.
Tizen uses SkiaSharp to implement PancakeView.

### Demo: Thewissen.PancakeView.Sample
pancakeview_tizen4

### Platform support
Missing features, `Borders.DrawindStyle` and `Shadows`, will soon be implemented and supported.
#### Borders
Property | iOS | Android | UWP | WPF | macOS | Tizen
-- | -- | -- | -- | -- | -- | --
Color | ✅ | ✅ | ✅ | ❌ | ❌ | ✅
DashPattern | ✅ | ✅ | ❌ | ❌ | ✅ | ✅
DrawingStyle | ✅ | ❌ | ❌ | ❌ | ✅ | ❌
GradientEndPoint | ✅ | ✅ | ✅ | ❌ | ✅ | ✅
GradientStartPoint | ✅ | ✅ | ✅ | ❌ | ✅ | ✅
GradientStops | ✅ | ✅ | ✅ | ❌ | ✅ | ✅
Thickness | ✅ | ✅ | ✅ | ❌ | ✅ | ✅

#### Gradients
Property | iOS | Android | UWP | WPF | macOS | Tizen
-- | -- | -- | -- | -- | -- | --
BackgroundGradientEndPoint | ✅ | ✅ | ✅ | ❌ | ✅ | ✅
BackgroundGradientStartPoint | ✅ | ✅ | ✅ | ❌ | ✅ | ✅
BackgroundGradientStops | ✅ | ✅ | ✅ | ❌ | ✅ | ✅

#### Rounded Corners
Property | iOS | Android | UWP | WPF | macOS | Tizen
-- | -- | -- | -- | -- | -- | --
CornerRadius | ✅ | ✅ | ✅ | ❌ | ✅ | ✅

#### Shadows
Property | iOS | Android | UWP | WPF | macOS | Tizen
-- | -- | -- | -- | -- | -- | --
BlurRadius | ✅ | ✅ | ✅ | ❌ | ✅ | ✅
Color | ✅ | ❌ | ✅ | ❌ | ✅ | ✅
Offset | ✅ | ❌ | ✅ | ❌ | ✅ | ✅
Opacity | ✅ | ❌ | ✅ | ❌ | ✅ | ✅

#### Shapes
Property | iOS | Android | UWP | WPF | macOS | Tizen
-- | -- | -- | -- | -- | -- | --
OffsetAngle | ✅ | ✅ | ❌ | ❌ | ✅ | ✅
Sides | ✅ | ✅ | ❌ | ❌ | ✅ | ✅

### CODE REVIEW

File: src/Xamarin.Forms.PancakeView.Multi/Platforms/Tizen/PancakeViewRenderer.cs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+                    _skCanvasView.PaintSurface -= OnPaintSurface;
+                    _skCanvasView.Unrealize();
+                    _skCanvasView = null;
+                }
+            }
+        }
+
+        protected override void UpdateBackgroundColor(bool initialize)
+        {
+        }
```
Comment (rookiejava) [inline src/Xamarin.Forms.PancakeView.Multi/Platforms/Tizen/PancakeViewRenderer.cs:69]: ?

File: example/Thewissen.PancakeViewSample.Tizen/Thewissen.PancakeViewSample.Tizen.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        {
+            base.OnCreate();
+            ElmSharp.Utility.AppendGlobalFontPath(this.DirectoryInfo.Resource);
+            LoadApplication(new App());
+        }
+
+        static void Main(string[] args)
+        {
+            var app = new Program();
+            Forms.Init(app);
```
Comment (rookiejava) [inline example/Thewissen.PancakeViewSample.Tizen/Thewissen.PancakeViewSample.Tizen.cs:18]:
```suggestion
            Forms.Init(app, true);
```

File: example/Thewissen.PancakeViewSample.Tizen/Thewissen.PancakeViewSample.Tizen.csproj [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+    <DebugType>portable</DebugType>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <DebugType>None</DebugType>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Folder Include="lib\" />
+  </ItemGroup>
```
Comment (rookiejava) [inline example/Thewissen.PancakeViewSample.Tizen/Thewissen.PancakeViewSample.Tizen.csproj:18]: Can omit these lines

File: example/Thewissen.PancakeViewSample.Tizen/tizen-manifest.xml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<manifest package="org.tizen.example.Thewissen.PancakeViewSample.Tizen" version="1.0.0" api-version="4" xmlns="http://tizen.org/ns/packages">
+    <profile name="common" />
+    <ui-application appid="org.tizen.example.Thewissen.PancakeViewSample.Tizen" exec="Thewissen.PancakeViewSample.Tizen.dll" multiple="false" nodisplay="false" taskmanage="true" type="dotnet" launch_mode="single">
+        <label>Thewissen.PancakeViewSample.Tizen</label>
+        <icon>Thewissen.PancakeViewSample.Tizen.png</icon>
+        <metadata key="http://tizen.org/metadata/prefer_dotnet_aot" value="false" />
```
Comment (rookiejava) [inline example/Thewissen.PancakeViewSample.Tizen/tizen-manifest.xml:7]: Why don't you use AOT?

File: src/Xamarin.Forms.PancakeView.Multi/Platforms/Tizen/PancakeViewRenderer.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+        protected override void OnElementPropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            base.OnElementPropertyChanged(sender, e);
+
+            var pancake = Element as PancakeView;
```
Comment (rookiejava) [inline src/Xamarin.Forms.PancakeView.Multi/Platforms/Tizen/PancakeViewRenderer.cs:75]: ??

**Comment by rookiejava** [comment]
I believe you can also implement `SHADOW` features marked as not supported. :-)

Property | iOS | Android | UWP | WPF | macOS | Tizen
-- | -- | -- | -- | -- | -- | --
BlurRadius | ✅ | ✅ | ✅ | ❌ | ✅ | ❌-> ✅
Color | ✅ | ❌ | ✅ | ❌ | ✅ | ❌-> ✅
Offset | ✅ | ❌ | ✅ | ❌ | ✅ | ❌-> ✅
Opacity | ✅ | ❌ | ✅ | ❌ | ✅ | ❌-> ✅

**Comment by sthewissen** [comment]
Let me start off by saying that it's awesome that you're PRing this in! Very cool!

I'll try to get a review up and running and see if I need to adjust something on the build pipeline side of things.

File: src/Xamarin.Forms.PancakeView.Multi/Platforms/Tizen/PancakeViewRenderer.cs (hunk 1 above)
Comment (JoonghyunCho) [reply src/Xamarin.Forms.PancakeView.Multi/Platforms/Tizen/PancakeViewRenderer.cs:69]: I intended to remove the base call of updating `BackgroundColor`, but found a better way. Fixed.

File: src/Xamarin.Forms.PancakeView.Multi/Platforms/Tizen/PancakeViewRenderer.cs (hunk 5 above)
Comment (JoonghyunCho) [reply src/Xamarin.Forms.PancakeView.Multi/Platforms/Tizen/PancakeViewRenderer.cs:75]: mistake...

**Comment by JoonghyunCho** [comment]
@sthewissen Huge thanks for welcoming my PR :) Glad you didn't freaked out.
I would just need to have little more time to make this complete and more stable.
I will also implement for `Shadow`.

**Comment by JoonghyunCho** [comment]
@rookiejava Absolutely. I was gonna update after implementing `Shadow` 😄

**Comment by sthewissen** [comment]
The CI/CD pipeline is crashing with the following error:

```
src\Xamarin.Forms.PancakeView.Multi\Platforms\Tizen\PancakeViewRenderer.cs#L38
src\Xamarin.Forms.PancakeView.Multi\Platforms\Tizen\PancakeViewRenderer.cs(38,45): Error CS0234: The type or namespace name 'NativeParent' does not exist in the namespace 'Xamarin.Forms' (are you missing an assembly reference?)
```

I'm assuming this has something to do with the version of Xamarin.Forms currently in use:
https://developer.samsung.com/tizen/blog/en-us/2019/09/02/changes-to-tizennet-from-xamarinforms-420

I personally have nothing against upgrading XamForms minimum version to something around 4.2 to support Tizen. Just let me know :)

**Comment by rookiejava** [comment]
> I personally have nothing against upgrading XamForms minimum version to something around 4.2 to support Tizen. Just let me know :)

Personally, it would be best if the minimum version can be 4.2 or higher. I understand that it is better to support more version, but isn't 3.6 too old? 😉

**Comment by JoonghyunCho** [comment]
@sthewissen Sounds awesome upgrading Xamarin.Forms version, but let me first try editing `Xamarin.Forms.PancakeView.Multi.csproj` so only `Tizen` can use the different version of Xamarin.Forms. This way could be better maybe? I will update it tomorrow :)

**Comment by sthewissen** [comment]
Just to fix the build pipeline I upgraded to 4.2 :) I'm fine with it as is.

**Comment by JoonghyunCho** [comment]
Thanks for updating the version. I am now actually trying to make just Tizen to use different version of Xamarin.Forms like `4.6.0.847`, and having difficulties of testing due to the following error.
>error MSB4064: The "ValidateOnly" parameter is not supported by the "XamlCTask" task.
>error MSB4063: The "XamlCTask" task could not be initialized with its input parameters. 
The error irregularly occurs when I build a sample application. it seems to be occurred due to the mismatch of Xamarin.Forms version in an application and library(not sure). The app sometimes builds OK, and sometimes not due to this error in my environment. Do you have any idea what I missed?

**Comment by sthewissen** [comment]
I've also found that this is exactly like you said, a problem in Xamarin.Forms & compiling locally. I have also noticed this usually appears when the multitargeting library runs a different XamForms version than the sample app. I have a commented out line in the multi-target project `csproj` file for development purposes to match these. I simply don't commit that line when doing the final build for the package :( Can't say I've found a better solution so far.

**Comment by JoonghyunCho** [comment]
It seems like the same issue here: https://github.com/xamarin/Xamarin.Forms/issues/8209

**Comment by sthewissen** [comment]
@JoonghyunCho Looks awesome! A few questions:

- Can I test this somehow myself? I have 0 experience working with Tizen 😅
- Is the work on your end done, so could it be potentially merged?

**Comment by JoonghyunCho** [comment]
@sthewissen Updating X.F 4.2 for PancakeView seems having potential issues (XamlCTask error on higher X.F version of application.)... so I've tried 1. reverting X.F version back to `3.6.0.220655` 2. adding condition for Tizen to use `4.4.0.991757`.

- You can definitely test this yourself. 
  - Check out the setup guide [here](https://docs.tizen.org/application/vstools/install/) and [Creating an application](https://developer.samsung.com/tizen/Galaxy-Watch/Quickstarts/Creating-an-application.html) is just to take a look.
  - Basically, you need to install an extension on Visual Studio and a Tizen baseline sdk with emulator.
    - Just to make it easier, you can try running `PancakeViewSample` on Mobile emulator.
    - 5.5 Mobile or 4.0 Mobile emulator is recommended. I mostly used 5.5 because it is the latest.
    - It definetely works on wearable emulator, but the layout for sample app breaks on the watch. 
- I think it's ok to be merged :) I can do some more tests and create another PR for any discovered issues.

Let me know for any difficulties of testing on your side!

**Comment by sthewissen** [comment]
@JoonghyunCho I tried testing and have my emulator and all set up (VSMac). However, when I run the sample app it doesn't actually launch the app itself. It does put it on the emulator, but after that the process seems to stop. I understand that this description doesn't tell you very much, but I also can't really find a lot of things going wrong in the Output logs etc. 🤔

**Comment by JoonghyunCho** [comment]
@sthewissen The app is installed but not launched? Could you right click(probably similar on Mac) the emulator and launch `Shell`?  Following command will show the related log including exception. I will check if you share the log.
```
sh-3.2$ dlogutil DOTNET_LAUNCHER
```
If there is System.BadImageFormatException, rebuilding PancakeView and sample app will help.

**Comment by sthewissen** [comment]
I do indeed get an error similar to that:

```
I/DOTNET_LAUNCHER( 3289): System.BadImageFormatException: Duplicate type with name 'Thewissen.PancakeViewSample.<>PropertyChangedEventArgs' in assembly 'Thewissen.PancakeViewSample, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
I/DOTNET_LAUNCHER( 3289):    at Thewissen.PancakeViewSample.Tizen.Program.OnCreate()
I/DOTNET_LAUNCHER( 3289):    at Tizen.Applications.CoreBackend.UICoreBackend.OnCreateNative(IntPtr data)
I/DOTNET_LAUNCHER( 3289):    at Tizen.Applications.CoreBackend.UICoreBackend.Run(String[] args)
I/DOTNET_LAUNCHER( 3289):    at Tizen.Applications.CoreApplication.Run(String[] args)
I/DOTNET_LAUNCHER( 3289):    at Tizen.Applications.CoreUIApplication.Run(String[] args)
I/DOTNET_LAUNCHER( 3289):    at Thewissen.PancakeViewSample.Tizen.Program.Main(String[] args) in /Users/sthewissen/Documents/GitHub/Xamarin.Forms.PancakeView/example/Thewissen.PancakeViewSample.Tizen/Thewissen.PancakeViewSample.Tizen.cs:line 22
I/DOTNET_LAUNCHER( 3289): onSigabrt called
```

Clean/rebuilds unfortunately don't seem to fix the problem. Either way, I'm not against merging this. You guys are the experts on Tizen so a problem on my local environment shouldn't stop this from being merged 👍

**Comment by JoonghyunCho** [comment]
@sthewissen Thanks a lot for having trust on us :) but still it's not good if anyone face the same issue running sample app on Tizen.
I updated Fody version to 6.2.1 which resolves `System.BadImageFormatException` on my side. Could you test again after maybe uninstall the app on emulator just to make sure? You can uninstall an app by long pressing the icon.

**Pull request merged**